ide/editor.completion/src/org/netbeans/spi/editor/completion/CompletionCollector.java

```diff
@@ -0,0 +1,565 @@
+             * top of the file if the completion item will insert an unqualified type).
+             */
+            @NonNull
+            public Builder additionalTextEdits(@NonNull List<TextEdit> additionalTextEdits) {
+                this.additionalTextEdits = additionalTextEdits;
+                return this;
+            }
+
+            /**
+             * A list of characters that when pressed while this completion is
+             * active will accept it first and then type that character. All
+             * commit characters should have length one and that superfluous
+             * characters will be ignored.
+             */
+            @NonNull
+            public Builder commitCharacters(@NonNull List<String> commitCharacters) {
+                this.commitCharacters = commitCharacters;
+                return this;
+            }
+
+            /**
+             * A command that is executed after inserting this completion.
+             */
+            @NonNull
+            public Builder command(@NonNull Command command) {
+                this.command = command;
+                return this;
+            }
+
+            /**
+             * A data entry to identify this completion.
+             */
+            @NonNull
+            public Builder data(@NonNull Object data) {
+                this.data = data;
+                return this;
+            }
+
+            /**
+             * Builds completion.
+             */
+            @NonNull
+            public Completion build() {
+                return new Completion(label, kind, tags, detail, documentation, preselect,
+                        sortText, filterText, insertText, insertTextFormat, textEdit,
+                        additionalTextEdits, commitCharacters, command, data);
+            }
+        }
+    }
+
+    public static enum Kind {
+
+        Text(1),
+	Method(2),
+	Function(3),
+	Constructor(4),
+	Field(5),
+	Variable(6),
+	Class(7),
+	Interface(8),
+	Module(9),
+	Property(10),
+	Unit(11),
+	Value(12),
+	Enum(13),
+	Keyword(14),
+	Snippet(15),
+	Color(16),
+	File(17),
+	Reference(18),
+	Folder(19),
+	EnumMember(20),
+	Constant(21),
+	Struct(22),
+	Event(23),
+	Operator(24),
+	TypeParameter(25);
+
+        private final int value;
+
+	private Kind(int value) {
+            this.value = value;
+	}
+
+	public int getValue() {
+	    return value;
+	}
+    }
+
+    public static enum Tag {
+
+        Deprecated(1);
+
+        private final int value;
+
+        private Tag(int value) {
+            this.value = value;
+        }
+
+        public int getValue() {
+            return value;
+        }
+    }
+
+    public static enum TextFormat {
+
+        /**
+         * The primary text to be inserted is treated as a plain string.
+         */
+        PlainText(1),
+
+        /**
+         * The primary text to be inserted is treated as a snippet.
+         */
+        Snippet(2);
```

Comment:
Can we only support Snippets in this API? One can simulate `PlainText` via `Snippet` with no parameters, right?

Reply:
You cannot send Snippets to a LSP client that does not support them (a capability to support Snippets is sent as part of `initialize` request). I would not hide that.